cpp/src/io/text/data_chunk_source_factories.cpp

```diff
@@ -0,0 +1,282 @@
+};
+
+/**
+ * @brief a reader which produces views of device memory which contain a copy of the data from an
+ * istream.
+ *
+ */
+class istream_data_chunk_reader : public data_chunk_reader {
+  struct host_ticket {
+    cudaEvent_t event;
+    thrust::host_vector<char, thrust::system::cuda::experimental::pinned_allocator<char>> buffer;
+  };
+
+ public:
+  istream_data_chunk_reader(std::unique_ptr<std::istream> datastream)
+    : _datastream(std::move(datastream)), _tickets(2)
+  {
+    // create an event to track the completion of the last device-to-host copy.
+    for (std::size_t i = 0; i < _tickets.size(); i++) {
+      CUDA_TRY(cudaEventCreate(&(_tickets[i].event)));
+    }
+  }
+
+  ~istream_data_chunk_reader()
+  {
+    for (std::size_t i = 0; i < _tickets.size(); i++) {
+      CUDA_TRY(cudaEventDestroy(_tickets[i].event));
+    }
+  }
+
+  void skip_bytes(std::size_t size) override { _datastream->ignore(size); };
+
+  std::unique_ptr<device_data_chunk> get_next_chunk(std::size_t read_size,
+                                                    rmm::cuda_stream_view stream) override
+  {
+    CUDF_FUNC_RANGE();
+
+    auto& h_ticket = _tickets[_next_ticket_idx];
+
+    _next_ticket_idx = (_next_ticket_idx + 1) % _tickets.size();
+
+    // synchronize on the last host-to-device copy, so we don't clobber the host buffer.
+    CUDA_TRY(cudaEventSynchronize(h_ticket.event));
+
+    // resize the host buffer as necessary to contain the requested number of bytes
+    if (h_ticket.buffer.size() < read_size) { h_ticket.buffer.resize(read_size); }
+
+    // read data from the host istream in to the pinned host memory buffer
+    _datastream->read(h_ticket.buffer.data(), read_size);
+
+    // adjust the read size to reflect how many bytes were actually read from the data stream
+    read_size = _datastream->gcount();
+
+    // get a view over some device memory we can use to buffer the read data on to device.
+    auto chunk = rmm::device_uvector<char>(read_size, stream);
+
+    // copy the host-pinned data on to device
+    CUDA_TRY(cudaMemcpyAsync(  //
+      chunk.data(),
+      h_ticket.buffer.data(),
+      read_size,
+      cudaMemcpyHostToDevice,
+      stream.value()));
+
+    // record the host-to-device copy.
+    CUDA_TRY(cudaEventRecord(h_ticket.event, stream.value()));
+
+    // return the view over device memory so it can be processed.
+    return std::make_unique<device_uvector_data_chunk>(std::move(chunk));
+  }
+
+ private:
+  std::size_t _next_ticket_idx = 0;
+  std::unique_ptr<std::istream> _datastream;
+  std::vector<host_ticket> _tickets;
+};
+
+class gds_data_chunk_reader : public data_chunk_reader {
+ public:
+  gds_data_chunk_reader(std::string filename) : _filename{filename}
+  {
+    device_reader = cudf::io::detail::make_cufile_input(_filename);
+    if (!device_reader) {
+      host_reader = std::make_unique<istream_data_chunk_reader>(
+        std::make_unique<std::ifstream>(_filename, std::ifstream::in));
+    }
```

Comment:
this can throw an exception if `device_reader` is null, and we can catch it and fall back to a different reader upstream.